Only in inetutils-1.9.4-anno/telnet: authenc.xml
diff -u -r inetutils-1.9.4/telnet/commands.c inetutils-1.9.4-anno/telnet/commands.c
--- inetutils-1.9.4/telnet/commands.c	2025-07-08 06:14:44.000000000 +0800
+++ inetutils-1.9.4-anno/telnet/commands.c	2025-09-24 15:34:51.639113000 +0800
@@ -92,7 +92,7 @@
 #include "xalloc.h"
 #include "xvasprintf.h"
 #include "xalloc.h"
-
+#include <valgrind/flowcheck.h>
 #ifdef	ENCRYPTION
 # include <libtelnet/encrypt.h>
 #endif
@@ -2661,7 +2661,7 @@
 
       printf ("Trying %s...\n", buf);
 
-      net = socket (aip->ai_family, SOCK_STREAM, 0);
+      net = socket (aip->ai_family, SOCK_STREAM, 0); FC_TAINT_WORD(&net);
       if (net < 0)
 	{
 	  perror ("telnet: socket");
@@ -2675,7 +2675,7 @@
 	    perror ("setsockopt (SO_DEBUG)");
 	}
 
-      err = connect (net, (struct sockaddr *) aip->ai_addr, aip->ai_addrlen);
+      err = connect (net, (struct sockaddr *) aip->ai_addr, aip->ai_addrlen); FC_TAINT_WORD(&err);
       if (err < 0)
 	{
 	  if (aip->ai_next)
@@ -2737,7 +2737,7 @@
   printf ("Trying %s...\n", inet_ntoa (sin.sin_addr));
   do
     {
-      net = socket (AF_INET, SOCK_STREAM, 0);
+      net = socket (AF_INET, SOCK_STREAM, 0); FC_TAINT_WORD(&net);
       if (net < 0)
 	{
 	  perror ("telnet: socket");
Only in inetutils-1.9.4-anno/telnet: commands.xml
Only in inetutils-1.9.4-anno/telnet: main.xml
Only in inetutils-1.9.4-anno/telnet: network.xml
Only in inetutils-1.9.4-anno/telnet: ring.xml
diff -u -r inetutils-1.9.4/telnet/sys_bsd.c inetutils-1.9.4-anno/telnet/sys_bsd.c
--- inetutils-1.9.4/telnet/sys_bsd.c	2015-03-31 23:40:50.000000000 +0800
+++ inetutils-1.9.4-anno/telnet/sys_bsd.c	2025-09-24 15:32:41.878385000 +0800
@@ -70,7 +70,7 @@
 #include "defines.h"
 #include "externs.h"
 #include "types.h"
-
+#include <valgrind/flowcheck.h>
 #ifdef	SIGINFO
 extern void ayt_status ();
 extern void sendayt ();
@@ -1229,10 +1229,10 @@
 	  ioctl (net, SIOCATMARK, (char *) &atmark);
 	  if (atmark)
 	    {
-	      c = recv (net, netiring.supply, canread, MSG_OOB);
+	      c = recv (net, netiring.supply, canread, MSG_OOB); int temp=c/4;for(int i=0;i<temp;i++){FC_TAINT_WORD(netiring.supply +4*i);}
 	      if ((c == -1) && (errno == EINVAL))
 		{
-		  c = recv (net, netiring.supply, canread, 0);
+		  c = recv (net, netiring.supply, canread, 0);int temp=c/4;for(int i=0;i<temp;i++){FC_TAINT_WORD(netiring.supply +4*i);}
 		  if (clocks.didnetreceive < clocks.gotDM)
 		    {
 		      SYNCHing = stilloob (net);
@@ -1252,7 +1252,7 @@
 		   * increment the count...)
 		   */
 		  int i;
-		  i = recv (net, netiring.supply + c, canread - c, MSG_OOB);
+		  i = recv (net, netiring.supply + c, canread - c, MSG_OOB); int temp=i/4;for(int i=0;i<temp;i++){FC_TAINT_WORD(netiring.supply +c +4*i);}
 		  if (i == c &&
 		      memcmp (netiring.supply, netiring.supply + c, i) == 0)
 		    {
@@ -1275,23 +1275,24 @@
 		   * to clear the atmark to get out of
 		   * an infinate loop.
 		   */
-		  i = read (net, netiring.supply + c, canread - c);
+		  i = read (net, netiring.supply + c, canread - c); int temp=i/4;for(int i=0;i<temp;i++){FC_TAINT_WORD(netiring.supply + c+4*i);}
+    }
 		  if (i > 0)
 		    c += i;
 		}
 	    }
 	  else
 	    {
-	      c = recv (net, netiring.supply, canread, 0);
+	      c = recv (net, netiring.supply, canread, 0); int temp=c/4;for(int i=0;i<temp;i++){FC_TAINT_WORD(netiring.supply +4*i);}
 	    }
 	}
       else
 	{
-	  c = recv (net, netiring.supply, canread, 0);
+	  c = recv (net, netiring.supply, canread, 0);int temp=c/4;for(int i=0;i<temp;i++){FC_TAINT_WORD(netiring.supply +4*i);}
 	}
       settimer (didnetreceive);
 #else /* !defined(SO_OOBINLINE) */
-      c = recv (net, (char *) netiring.supply, canread, 0);
+      c = recv (net, (char *) netiring.supply, canread, 0);int temp=c/4;for(int i=0;i<temp;i++){FC_TAINT_WORD(netiring.supply +4*i);}
 #endif /* !defined(SO_OOBINLINE) */
       if (c < 0 && errno == EWOULDBLOCK)
 	{
Only in inetutils-1.9.4-anno/telnet: sys_bsd.xml
Only in inetutils-1.9.4-anno/telnet: telnet.xml
Only in inetutils-1.9.4-anno/telnet: terminal.xml
Only in inetutils-1.9.4-anno/telnet: tn3270.xml
Only in inetutils-1.9.4-anno/telnet: utilities.xml
diff -u -r inetutils-1.9.4/.vscode/settings.json inetutils-1.9.4-anno/.vscode/settings.json
--- inetutils-1.9.4/.vscode/settings.json	2025-07-19 15:13:49.000000000 +0800
+++ inetutils-1.9.4-anno/.vscode/settings.json	2025-09-24 14:37:23.187458556 +0800
@@ -21,6 +21,7 @@
         "type_traits": "c",
         "typeinfo": "c",
         "numeric": "c",
-        "algorithm": "c"
+        "algorithm": "c",
+        "stdarg.h": "c"
     }
 }
\ No newline at end of file
